### DISCUSSION TITLE
Loading in new excel based tables

### DISCUSSION THREAD

**Discussion by dogmom32** [lifelib-dev/lifelib#44] [General]
Hello, 

I'm somewhat new to this LifeLib and ModelX but where/how in the `__init__.py` code for the BasicTerm_S model are the reference tables loaded in? I see that it uses the "_mx_dataclient" but would like to understand more about this.

**Comment by fumitoh** [comment]
`_mx_dataclient` is not used in the latest version anymore. The doc should be updated. It was used to hold the metadata of referenced objects, but now the metadata is stored in the model.

By loading BasicTerm_S into memory by `read_model` like this,

```python
>>> model = mx.read_model("C:\BasicTerm_S")
```
`model.Projection.model_point_table` and `model.Projection.mort_table` for example refer to DataFrames read from Excel files in the model directory. These were origially assigned using new_pandas method. The meta data of the DataFrames, such as file path and file type are stored in the model as PandasData objects.
You can list the PandasData objects by `model.iospecs`

```python
>>> BasicTerm_S.iospecs
[<PandasData path='disc_rate_ann.xlsx' file_type='excel'>,
 <PandasData path='model_point_table.xlsx' file_type='excel'>,
 <PandasData path='mort_table.xlsx' file_type='excel'>]
```
`get_spec` returns the associated PandasData for a referenced object:

```python
>>> model.get_spec(model.Projection.model_point_table)
<PandasData path='model_point_table.xlsx' file_type='excel'>
```

The tutorial page below explains the mechanism.

https://docs.modelx.io/en/latest/tutorial/using_pandas.html

**Reply by dogmom32** [reply]
How would go about repurposing the Basic Term S model by changing the inforce file and mortality table so the model reads in the updated information? 
I read through the documentation of new pandas and the timing of when the dataframe is created, then loaded into the model by "new_pandas" and then the model is saved. 

Basically, how would I go about updating the code to take in new inputs?

**Reply by fumitoh** [reply]
To use your own inforce file, you can just replace model_point_table.xlsx in the directory with your file.
Then load the model by `read_model` and `model.Projection.model_point_table` should refer to a DataFrame created from the updated file.
Another way is to remove the original DataFrame, load your data as DataFrame and assign it to `model_point_table` using `new_pandas`.

To change formulas that refer to the DataFrame, you can just update the formulas by `defcells`.
For example, if your inforce file contains a column for premium per policy named `prem_pp` and you want to use the column's values instead of calculating the premium per policy from loadings and net premium per policy, you can update the `premium_pp` cells like this.

```python
>>> model.Projection.premium_pp.formula   # Show original formula
def premium_pp():
    """Monthly premium per policy

    Monthly premium amount per policy defined as::

        round((1 + loading_prem()) * net_premium(), 2)

    .. versionchanged:: 0.2.0
       The ``t`` parameter is removed.

    .. seealso::

        * :func:`loading_prem`
        * :func:`net_premium_pp`

    """
    return round((1 + loading_prem()) * net_premium_pp(), 2)

>>> proj = model.Projection  # define alias for convenience

>>> @mx.defcells(proj)    # Update the formula of premium_pp. Give the parent space of premium_pp as argument to the decorator
... def premium_pp():
       return model_point()["prem_pp"]   
```
Another way to do this is

```python
>>> def temp():   # Define as a Python function. The name doesn't matter.
...     return model_point()["prem_pp"]   

>>> proj.premium_pp.formula = temp  # Then assign the function to the formula of premium_pp. 

>>> proj.promum_pp.formula   # Check the formula is updated
def premium_pp():
    return model_point()["prem_pp"]

```
Or, you can use Spyder with the spyder-modelx plugn, which I think is the easiest.

**Reply by fumitoh** [reply]
A completely different approch can be taken, which is to save the path to an input file in a model, and read the file from within a cells.

```python
>>> proj.file_path = r"C:\path\to\your\file.xlsx"      # Assign your file path as a string

>>> @mx.defcells(proj)
... def another_table():    # Create a DataFrame from the file
...     return pd.read_excel(file_path, index_col=0)

>>> @mx.defcells(proj)   # Update model_point to pick a model point from another_table()
... def model_point():
...     return another_table().loc[point_id]    # Need ()
```

**Comment by dogmom32** [comment]
Hey sorry about this but I'm having a hard time saving objects, then writing the model to a new folder and not seeing the saved object.  Here are my steps:
1) read existing model (e.g. BaseTerm_SE)
2) load in inforce (e.g. temp_inforce)
 `temp_inforce =pd.read("inforce.csv")`
3) follow pandas object instructions by setting the space.model_point_table = temp_inforce
`space.model_point_table = temp_inforce`
4) save table in model space directory
`space.new_pandas("model_point_table",` "space/model_point_table.xlsx", data=model_point_table, `file_type="excel")`
5) write model to a new location
6) read model from step 5 ; and I still don't see the newly saved inforce file

**Reply by fumitoh** [reply]
I think what you should do is:


```python
>>> temp_inforce =pd.read("inforce.csv")

>>> del space.model_point_table # delete the existing DataFrame

>>> space.new_pandas("model_point_table", "space/model_point_table.xlsx", data=temp_inforce , file_type="excel")    # or file_type="csv"

```
Your 3rd step is not needed. 

Note that the 3rd line above saves `temp_inforce` as "model_point_table.xlsx" within a subfolder named "space" in the BaseTerm_SE folder. If you mean to save the file under the Projection folder, then the second parameter should be "Projection/model_point_table.xlsx".

**Reply by dogmom32** [reply]
My ultimate goal is to build a model with update input files and have it where I can save or write it out to a new version with all pandas objects (inforce, assumption tables, etc) saved (similar to how your models are saved/stored).
